internal/compare/compare.go

```diff
@@ -0,0 +1,141 @@
+// Copyright (c) 2021 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package compare
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"go.uber.org/thriftrw/compile"
+)
+
+// Diagnostic is a message associated with an error and a file name.
+type Diagnostic struct {
+	File    string // File where error was discovered
+	Message string // Message contains error message.
+}
+
+func (d *Diagnostic) String() string {
+	return fmt.Sprintf("%s:%s", d.File, d.Message)
+}
+
+// Pass provides all reported errors.
+type Pass struct {
+	lints []Diagnostic
+}
+
+// Report reports an error.
+func (p *Pass) Report(d Diagnostic) {
+	p.lints = append(p.lints, d)
+}
+
+// Lints returns all errors.
+func (p *Pass) Lints() []Diagnostic {
+	return p.lints
+}
+
+func (p *Pass) String() string {
+	var b strings.Builder
+	for _, l := range p.lints {
+		_, _ = fmt.Fprintf(&b, "%s\n", l.String())
+	}
+
+	return b.String()
+}
+
+// Modules looks for removed methods and added required fields.
+func (p *Pass) Modules(from, to *compile.Module) {
+	for name, fromService := range from.Services {
+		p.service(fromService, to.Services[name])
+	}
+
+	for n, fromType := range from.Types {
+		p.typ(fromType, to.Types[n], filepath.Base(from.ThriftPath))
```

Comment:
I think `to.Types[n]` can be nil here so we'd wanna add a nil check at top of `func (p *Pass) typ` or here

Reply:
if there's no key, then value will be the map's zero-value, `TypeSpec`, not nil which will then get handled in `typ`. if you are looking at other cases where we handle it, it's because for those maps the values are pointers. 

from flow point of view, if we dont have the type, in this version, we can't handle renames yet, so we aren't going to error. renaming a struct is backwards compatible.

Reply:
So if `to.Types[n]` returns nil because the value doesn't exist in the `to`, the following won't `ok`:

    t, ok := to.(*compile.StructSpec)

It will refuse to turn a `nil` `TypeSpec` into a `*compile.StructSpec`, and return `false`.
So the nil check is kinda already done.